arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/continuations/EffectScope.kt

```diff
@@ -212,6 +214,55 @@ public interface EffectScope<in R> {
    */
   public suspend fun ensure(condition: Boolean, shift: () -> R): Unit =
     if (condition) Unit else shift(shift())
+
+  /**
+   * Encloses an action for which you want to catch any `shift`.
+   * [attempt] is used in combination with [catch].
+   *
+   * ```
+   * attempt { ... } catch { ... }
+   * ```
+   *
+   * The [f] may `shift` into a different `EffectScope`, giving
+   * the chance for a later [catch] to change the shifted value.
+   * This is useful to simulate re-throwing of exceptions.
+   */
+  @OptIn(ExperimentalTypeInference::class)
+  public suspend fun <E, A> attempt(
+    @BuilderInference
+    f: suspend EffectScope<E>.() -> A,
+  ): suspend EffectScope<E>.() -> A = f
+
+  /**
+   * When the [Effect] has shifted with [R] it will [recover]
+   * the shifted value to [A], and when it ran the computation to
+   * completion it will return the value [A].
+   * [catch] is used in combination with [attempt].
+   *
+   * ```kotlin
+   * import arrow.core.Either
+   * import arrow.core.None
+   * import arrow.core.Option
+   * import arrow.core.Validated
+   * import arrow.core.continuations.effect
+   * import io.kotest.assertions.fail
+   * import io.kotest.matchers.shouldBe
+   *
+   * suspend fun main() {
+   *   effect<String, Int> {
+   *     val x = Either.Right(1).bind()
+   *     val y = Validated.Valid(2).bind()
+   *     val z =
+   *      attempt { None.bind { "Option was empty" } } catch { 0 }
+   *     x + y + z
+   *   }.fold({ fail("Shift can never be the result") }, { it shouldBe 3 })
+   * }
+   * ```
+   * <!--- KNIT example-effect-scope-09.kt -->
+   */
+  public suspend infix fun <E, A> (suspend EffectScope<E>.() -> A).catch(
+    recover: EffectScope<R>.(E) -> A,
```

Comment:
Should this not be `recover: suspend EffectScope<R>.(E) -> A`? My eye only fell on this just now.

Reply:
No, the handler executes in the same context available to the outer thing (which is called R here). The idea is that you “close over” errors of type E, and may turn them into errors of type R, if that makes sense.

Reply:
Yes, makes sense but not sure I understand why suspend is not allowed in recover 🤔

Reply:
Doesn’t this prevent you from calling shift or bind?

Reply:
Oh, I forgot it when I moved it from being an extension function! I’ll push an update tonight.

Reply:
Awesome, feel free to merge after! Great contribution, excited to see this getting merged 🥳